core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java

```diff
@@ -311,6 +314,10 @@ public RelNode getRoot() {
     return ImmutableList.copyOf(materializations);
   }
 
+  @Override public void registerMaterializationRules(List<UnifyRule> rules) {
+    materializationRules = rules;
```

Comment:
how about make `materializationRules` immutable, and use `ImmutableList.copyOf` here?

Reply:
hmm, why materializationRules should be immutable ?
Regarding the test case, I will add one soon ~